plugins/inputs/proxmox/proxmox.go

```diff
@@ -0,0 +1,259 @@
+  response_timeout = "5s"
+`
+
+func (px *Proxmox) SampleConfig() string {
+	return sampleConfig
+}
+
+func (px *Proxmox) Description() string {
+	return "Provides metrics from Proxmox nodes (Proxmox Virtual Environment > 6.2)."
+}
+
+func (px *Proxmox) Gather(acc telegraf.Accumulator) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+	px.hostname = hostname
+
+	err = getNodeSearchDomain(px)
+	if err != nil {
+		return err
+	}
+
+	gatherLxcData(px, acc)
+	gatherQemuData(px, acc)
+
+	return nil
+}
+
+func init() {
+	px := Proxmox{
+		requestFunction: performRequest,
+	}
+	inputs.Add("proxmox", func() telegraf.Input { return &px })
+}
+
+func getNodeSearchDomain(px *Proxmox) error {
+	apiUrl := "/nodes/" + px.hostname + "/dns"
+	jsonData, err := px.requestFunction(px, apiUrl, http.MethodGet, nil)
+
+	var nodeDns NodeDns
+	err = json.Unmarshal(jsonData, &nodeDns)
+	if err != nil {
+		return err
+	}
+	px.nodeSearchDomain = nodeDns.Data.Searchdomain
+
+	return nil
+}
+
+func performRequest(px *Proxmox, apiUrl string, method string, data url.Values) ([]byte, error) {
+	tlsCfg, err := px.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, err
+	}
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+		},
+		Timeout: px.ResponseTimeout.Duration,
+	}
+
+	request, err := http.NewRequest(method, px.BaseUrl+apiUrl, strings.NewReader(data.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add("Authorization", "PVEAPIToken="+px.ApiToken)
+
+	resp, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return responseBody, nil
+}
+
+func gatherLxcData(px *Proxmox, acc telegraf.Accumulator) {
+	gatherVmData(px, acc, true)
+}
+
+func gatherQemuData(px *Proxmox, acc telegraf.Accumulator) {
+	gatherVmData(px, acc, false)
+}
+
+func gatherVmData(px *Proxmox, acc telegraf.Accumulator, lxc bool) {
+	vmStats, err := getVmStats(px, lxc)
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+
+	// For each VM add metrics to Accumulator
+	for _, vmStat := range vmStats.Data {
+		vmConfig, err := getVmConfig(px, vmStat.Id, lxc)
+		if err != nil {
+			log.Fatal(err)
+			return
+		}
+		tags := getTags(px, vmStat.Name, vmConfig, lxc)
+		fields, err := getFields(vmStat)
+		if err != nil {
+			log.Fatal(err)
+			return
+		}
+		acc.AddFields("proxmox", fields, tags)
+	}
+}
+
+func getVmStats(px *Proxmox, lxc bool) (VmStats, error) {
+	vmPath := "/qemu"
+	if lxc {
+		vmPath = "/lxc"
+	}
+	apiUrl := "/nodes/" + px.hostname + vmPath
+	jsonData, err := px.requestFunction(px, apiUrl, http.MethodGet, nil)
+	if err != nil {
+		return VmStats{}, err
+	}
+
+	var vmStats VmStats
+	err = json.Unmarshal(jsonData, &vmStats)
+	if err != nil {
+		return VmStats{}, err
+	}
+
+	return vmStats, nil
+}
+
+func getVmConfig(px *Proxmox, vmId string, lxc bool) (VmConfig, error) {
```

Comment:
not a big fan of boolean args. a resource type might be better?
```
type resourceType string
var (
  rtQEMU resourceType = "qemu"
  rtLXC resourceType = "lxc"
)
```
This would let you make the `lxc bool` arg just `rt resourceType` and include it directly in the path.